versiondb/client/changeset.go

```diff
@@ -0,0 +1,322 @@
+	ZlibFileSuffix   = ".zz"
+	SnappyFileSuffix = ".snappy"
+)
+
+// WriteChangeSet writes a version of change sets writer.
+//
+// Change set file format:
+// ```
+// version: int64
+// size: int64         // size of whole payload
+// payload:
+//   delete: int8
+//   keyLen: int16
+//   key
+//   [                 // if delete is false
+//     valueLen: int32
+//     value
+//   ]
+//
+// repeat with next version
+// ```
+func WriteChangeSet(writer io.Writer, version int64, cs iavl.ChangeSet) error {
+	var size int
+	items := make([][]byte, 0, len(cs.Pairs))
+	for _, pair := range cs.Pairs {
+		buf, err := encodeKVPair(pair)
+		if err != nil {
+			return err
+		}
+		size += len(buf)
+		items = append(items, buf)
+	}
+
+	var versionHeader [16]byte
+	binary.LittleEndian.PutUint64(versionHeader[:], uint64(version))
+	binary.LittleEndian.PutUint64(versionHeader[8:], uint64(size))
+
+	if _, err := writer.Write(versionHeader[:]); err != nil {
+		return err
+	}
+	for _, item := range items {
+		if _, err := writer.Write(item); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// ReadChangeSet decode a version of change set from reader.
+// if parseChangeset is false, it'll skip change set payload directly.
+//
+// returns (version, number of bytes read, changeSet, err)
+func ReadChangeSet(reader io.Reader, parseChangeset bool) (int64, int64, *iavl.ChangeSet, error) {
+	var versionHeader [16]byte
+	if _, err := io.ReadFull(reader, versionHeader[:]); err != nil {
+		return 0, 0, nil, err
+	}
+	version := binary.LittleEndian.Uint64(versionHeader[:8])
+	size := int64(binary.LittleEndian.Uint64(versionHeader[8:16]))
+	var changeSet iavl.ChangeSet
+	if size > 0 {
+		if parseChangeset {
+			limitedReader := io.LimitReader(reader, size)
+			var offset int64
+			for offset < size {
+				pair, err := readKVPair(limitedReader)
+				if err != nil {
+					return 0, 0, nil, err
+				}
+				offset += int64(encodedSizeOfKVPair(*pair))
+				changeSet.Pairs = append(changeSet.Pairs, *pair)
+			}
+		} else {
+			if _, err := io.CopyN(io.Discard, reader, size); err != nil {
+				return 0, 0, nil, err
+			}
+		}
+	}
+	return int64(version), size + 16, &changeSet, nil
+}
+
+// encodedSizeOfKVPair returns the encoded length of a key-value pair
+//
+// layout: deletion(1) + keyLen(2) + key + [ valueLen(4) + value ]
+func encodedSizeOfKVPair(pair iavl.KVPair) int {
+	size := 3 + len(pair.Key)
+	if !pair.Delete {
+		size += 4 + len(pair.Value)
+	}
+	return size
+}
+
+// encodeKVPair encode a key-value pair in change set.
+// see godoc of `encodedSizeOfKVPair` for layout description,
+// returns error if key/value length overflows.
+func encodeKVPair(pair iavl.KVPair) ([]byte, error) {
+	buf := make([]byte, encodedSizeOfKVPair(pair))
+
+	if len(pair.Key) > math.MaxUint16 {
+		return nil, fmt.Errorf("key length %d is too big", len(pair.Key))
+	}
+
+	binary.LittleEndian.PutUint16(buf[1:], uint16(len(pair.Key)))
+	copy(buf[3:], pair.Key)
+	if pair.Delete {
+		buf[0] = 1
```

Comment:
maybe do return and then remove the else condition?

Reply:
done